### PR TITLE
#159521695 Filter assets by model numbers and asset types

### PR DESCRIPTION
**Pull request by kimotho-njoki**
## What does this PR do?
Enables filtering of assets by model numbers and asset types.

## Description of Task to be completed?
- Create a filter component
- Refactor components to add filter functionality
- Add relevant tests

## How should this be manually tested?
- On the `Asset List` table, click on the `filter` button on the far right. Scroll and choose through the checkbox options to filter.

## What are the relevant pivotal tracker stories?
[#159521695](https://www.pivotaltracker.com/n/projects/2146417/stories/159521695)

## Any background context you want to add?
N/A

## Important notes
N/A

## Packages installed
* uuid

## Deployment note
N/A

## Related PRs branch | PR (branch_name) | (pr_link)
Branch | PR
------ | ------
other_pr_production | [link]()
other_pr_master | [link]()
## Todos
- [x] Raise PR
- [ ] Test
- [ ] Documentation
## Screenshots
- When filter button is `unclicked`
screen shot 2018-08-31 at 13 08 37

- When filter button is `clicked`
screen shot 2018-08-31 at 13 10 30

- When filtering by `asset types (headsets)`
screen shot 2018-08-31 at 13 09 58

### CODE REVIEW

File: src/components/common/CheckboxComponent.jsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+class CheckboxComponent extends Component {
+  state = {
+    isChecked: false
+  };
+
+  toggleCheckboxChange = () => {
+    const { handleCheckboxChange, label, name } = this.props;
+
+    this.setState(({ isChecked }) => (
```
Comment (sachungo) [inline src/components/common/CheckboxComponent.jsx]: Restructure to
```
  this.setState(({ isChecked }) => ({
     isChecked: !isChecked
  }));

File: src/components/common/FilterButton.jsx [hunk 2]
```diff
@@ -1,45 +1,49 @@
...
   }
 }
+
 FilterButton.propTypes = {
   render: PropTypes.func
 };
 
 FilterButton.defaultProps = {
-  render: () => {}
+  render: () => {
```
Comment (sachungo) [inline src/components/common/FilterButton.jsx]: Please revert to the previous version. I.e 
```
render: () => {}
```

**Comment by sachungo** [comment]
@kimotho-njoki @Arthur236, not all my comments have been addressed/answered, and....you missed some in the hidden conversation

image 2018-09-04 at 6 48 50 am

File: src/components/common/FilterButton.jsx [hunk 3]
```diff
@@ -42,8 +42,7 @@ FilterButton.propTypes = {
 };
 
 FilterButton.defaultProps = {
-  render: () => {
-  }
+  render: () => { }
```
Comment (sachungo) [inline src/components/common/FilterButton.jsx]: Why the empty space between `{}`?

File: src/components/Assets/AssetFilterComponent.jsx [hunk 4]
```diff
@@ -58,49 +58,50 @@ class AssetFilterComponent extends React.Component {
...
-            buttonName="apply filters"
-            color="primary"
-            handleClick={this.handleFilter}
-          />
-        </div> :
-        null
+      <React.Fragment>
+        <Accordion as={Menu} vertical className="filter-menu">
+          {
+            !isEmpty(options) ?
```
Comment (sachungo) [inline src/components/Assets/AssetFilterComponent.jsx]: I missed this yesterday - please change the order of the ternary operator to improve readability. I.e.
```
{isEmpty(options)
  ? <span>Loading filters...</span>
  : (THE_OTHER_TERNARY_OPTION)
}

File: src/_css/AssetDescriptionComponent.css [hunk 5]
```diff
@@ -1,42 +1,251 @@
+@font-face {
```
[the remark this answers is not in the document]
Comment (Arthur236) [reply src/_css/AssetDescriptionComponent.css]: This must have been added when we rebased. We'll correct it

**Comment by kimotho-njoki** [comment]
@sachungo we were in the process of working on your comments but the tests refactoring brought in a couple of bugs which we have just fixed

File: src/_css/FilterUserComponent.scss [hunk 6]
```diff
@@ -0,0 +1,39 @@
+@import "sharedStyling.scss";
+
+#filter-container {
```
[the remark this answers is not in the document]
Comment (kimotho-njoki) [reply src/_css/FilterUserComponent.scss]: @sachungo All we did was rename this file, it was not ours. Should we still make changes to it?

File: src/_actions/assets.action.js [hunk 7]
```diff
@@ -7,11 +7,19 @@ const {
   LOAD_ASSETS_STARTS
 } = constants;
 
+let url;
```
Comment (sachungo) [inline src/_actions/assets.action.js]: This should be declared in `getAssetsAction`

File: src/_actions/assets.action.js [hunk 8]
```diff
@@ -7,11 +7,19 @@ const {
...
+
 /* eslint-disable import/prefer-default-export */
-export const getAssetsAction = (pageNumber, limit) => (
-  (dispatch) => {
+export const getAssetsAction = (pageNumber, limit, filters) => {
+  if (filters) {
```
Comment (sachungo) [inline src/_actions/assets.action.js]: The `if else` can be simplified to:
```
let url = `manage-assets?page=${pageNumber}&page_size=${limit}`;
if (filters) {
  url = `manage-assets?page=${pageNumber}&page_size=${limit}&asset_type=${filters['Asset Types'] || ''}&model_number=${filters['Model Numbers'] || ''}`;
}

File: src/_css/FilterUserComponent.scss (hunk 6 above)
[the remark this answers is not in the document]
Comment (sachungo) [reply src/_css/FilterUserComponent.scss]: @kimotho-njoki If you just renamed the file, then ignore the comments related to it

**Comment by kimotho-njoki** [comment]
@sachungo All changes made... Merging now 😄 🍾

**Pull request merged**